doc: add design doc (#26)
nervosnetwork/toCKB: docs/design.md

# toCKB: An Asset Cross-Chain System

## Abstract

The toCKB is an asset cross-chain system, which consists of a bunch of protocols, CKB contracts and off-chain modules, to support decentralized, redeemable token on CKB supply-pegged to any asset on any blockchain.

> As long as you can transfer an asset on a blockchain system and construct a spv proof to verify the transaction, you can implement the spv verification logic in the toCKB contract. Then anyone can use the contract to mint a token on CKB pegged to the asset by locking the asset. They can transfer the token and burn it to redeem the asset on original blockchain as well. The peg is implemented using an approch mixed with bonding and spv proof verification, which involves a new role called signer. Signers act as a guard on the original blockchain to ensure the asset is locked when minting token and send the asset back to user when redeeming. Anyone can be signer by bonding an amount of CKB that ensures their behavior in the system remains honest, at risk of losing their bond in case of dishonesty or undercollateralization.

Reviewer: ```suggestion
As long as you can transfer an asset on a blockchain system and construct a spv proof to verify the transaction, you can implement the spv verification logic in the toCKB contract. Then anyone can use the contract to mint a token on CKB pegged to the asset by locking the asset. They can transfer the token and burn it to redeem the asset on original blockchain as well. The peg is implemented using an approch mixed with bonding and spv proof verification, which involves a new role called signer. Signers act as a guard on the original blockchain to ensure the asset is locked when minting token and send the asset back to user when redeeming. Anyone can be signer by bonding an amount of CKB that ensures their behavior in the system remains honest, at risk of losing their bond in case of dishonesty or inaction.
```

> The CKB contracts mediates the deposit's lifecycle, including request, bonding, signer fraud and redemption. The off-chain modules help users use the whole system painlessly, including constructing transactions, monitoring CKB as well as other blockchains to act automatically, generating spv proof and so on.

Reviewer: ```suggestion
The CKB contracts mediates the cross-chain's lifecycle, including request, bonding,  redemption and failure-handle. The off-chain modules help users use the whole system painlessly, including constructing transactions, monitoring CKB as well as other blockchains to act automatically, generating spv proof and so on.
```

- There will be additional logics in toCKB contract to handle unexpected situations, e.g. signer incentive, signer fraud, undercollateralization caused by the exchange rate changes.

## Process

The normal process shows below:

![CleanShot-20200908-230936-YL5UdBP8@2x](https://i.imgur.com/kMPOOOi.png)

<!--
```plantuml
autonumber
actor user
actor signer
database CKB
database XChain

== Mint XToken ==
user -> CKB: deposit request
signer -> CKB: bonding, provide a XChain receive addr
user -> XChain: transfer asset to signer
user -> CKB: use the proof of previous step to mint XToken
user -> user: user can use the XToken

== Redeem XAsset ==
user -> CKB: redeem request
signer -> XChain: signer send asset back to user
signer -> CKB: Withdraw Collateral
```
-->

1. User makes a deposit request on CKB, along with some pledge. If someone bonds as signer but the user does not follw up, he will lose the pledge to compensate the loss of signer's CKB liquidity.
2. Someone bonds CKB to become a signer, provide a XChain address for user to deposit XAsset.
3. User transfers their XAsset to signer on XChain.

> 4. User generates the transaction proof and relays it to CKB, mints 1-to-1 CKB token -- XToken.

Reviewer: miss signer-fee

5. User can deal with the XToken as he wish.
6. When user want to redeem XAsset back on XChain, he can make a redeem request on CKB, burn his XToken on CKB via toCKB contract.
7. Signer sends the XAsset back to user on XChain.
8. Signer generates the transaction proof, relay it on CKB, and withdraw his collateral back.

### Handle Failture

There are two kinds of failure: abort and faulty.

Abort:
1. After users made a deposit request, nobody bond their CKB to become the signer(step 2). Users can withdraw there pledge then.
2. After some bond CKB to become the signer, user does not send the XAsset to signer(step 3). Signer can withdraw his collateral and user's pledge.

> 3. During redeem process, signer does not send asset back to user(step 7). It will trigger liquidation.

Reviewer: ```suggestion
3. During redeem process, signer does not send asset back to user within a limited time(step 7). Anyone can trigger liquidation.
```

> 4. The exchange rate changes, causes the CKB value bonded lower than expected. It will trigger liquidation.

Reviewer: ```suggestion
4. The exchange rate changes, causes the CKB value bonded lower than expected. Anyone can trigger liquidation.
```

- Deposit Request
- Bonding
- Withdraw Pledge
- Withdraw Pledge and Collateral
- Pre-term Redeem
- At-term Redeem
- Withdraw Collateral
- Liquidation：SignerTimeout
- Redeem: Pre-Undercollateral
- Liquidation：Undercollateral
- Liquidation: FaultyWhenWarranty
- Liquidation: FaultyWhenRedeeming
- Auction: SignerTimeout
- Auction: FaultyWhenWarranty
- Auction: FaultyWhenRedeeming

#### Price Oracle

The toCKB system relies on a price oracle who provides the CKB/XAsset price to remain security.

In this stage, we will implement a centralized oracle which gets price from public APIs, and triggers update on specific conditions. It will be a little bit similar with [MakerDAO's price feed](https://developer.makerdao.com/feeds/).

We will improve it in the future, which may involve some decentralized oracle implemented on CKB, or even [reuse the data on Ethereum](https://talk.nervos.org/t/hi-ckb/4890).

#### The X Specific SPV Verification Logic

For different XChain, we have to implement the spv verification logic on CKB. Due to the flexible design of CKB-VM, it will be easy to reuse a lot of previous work.

### Off-Chain

> In this stage, we will make some scripts or command line tools to help users and signers construct transactions, monitor the chain events and even trigger the follow up actions automatically.

Reviewer: for off-chain, we will offer
- user-client
- signer-client
- guard-daemon

### Core Concepts

The core concepts of toCKB are:
- Because we want to support the chains lack of ability to verify CKB spv proof, we involve a new role called signer to replace the contract in spv based bridge. Users transfer their asset to signer's address instead of the contract to lock.
- Anyone can be signer by bonding an amount of CKB that ensures their behavior in the system remains honest. The exchange rate may changes over time, so the value of CKB have to be more than the locked asset for security. The collateral value percent is configurable, which is 150% in this stage.
- With the help of signer, we can handle both mint and redeem logic on CKB by spv proof verification:
    - When user locked their XAsset on XChain(which means they transfered their XAsset to signer), toCKB contract can verify the proof and mint XToken on CKB, owned by user.
    - When user burns their XToken on CKB, signer has to send the XAsset to user on XChain and submit the proof to CKB. toCKB contract verifies the proof, and redeem the bond of signers.

> - There will be additional logics in toCKB contract to handle unexpected situations, e.g. signer incentive, signer fraud, undercollateralization caused by the exchange rate changes.

Author: This part is not decided yet.
